File: sdks/typescript/src/core/canonicalize.ts
/**
 * DCP v2.0 Canonicalization — profile `dcp-jcs-v1`.
 *
 * Strict subset of RFC 8785 (JCS) with the cases JCS leaves
 * implementation-defined explicitly pinned. See
 * `spec/CANONICALIZATION_PROFILE.md` for the normative reference.
 *
 * Numeric rule: a number is valid iff its value is a finite integer.
 * `1.0`, `1.00`, and `1e2` are accepted because, post-parse, JavaScript
 * cannot distinguish them from `1` and `100`. `0.1`, `NaN`, and
 * infinities are rejected.
 *
 * Undefined handling (TypeScript-only): in objects, a key whose value
 * is `undefined` is omitted entirely; inside an array, `undefined`
 * serialises as the literal `null`. Other SDKs (Python, Rust, Go) do
 * not expose `undefined` as a distinct value.
 */

/**
 * Validate that a value contains no floating-point numbers anywhere in the tree.
 * Throws if a float is detected.
 */
export function assertNoFloats(value: unknown, path = '$'): void {
  if (value === null || value === undefined) return;
  if (typeof value === 'number') {
    if (!Number.isInteger(value)) {
      throw new Error(
        `Float value prohibited in DCP v2.0 at ${path}: ${value}. Use integer (e.g. millirisk 0-1000).`,
      );
    }
    return;
  }
  if (Array.isArray(value)) {
    for (let i = 0; i < value.length; i++) {
      assertNoFloats(value[i], `${path}[${i}]`);
    }
    return;
  }
  if (typeof value === 'object') {
    for (const [k, v] of Object.entries(value as Record<string, unknown>)) {
      assertNoFloats(v, `${path}.${k}`);
    }
  }
}

/**
 * RFC 8785 (JCS) canonical JSON serialization with DCP v2.0 restrictions:
 * - Object keys sorted lexicographically by Unicode code point
 * - Compact form (no whitespace)
 * - Integers only (floats rejected)
 * - null, true, false are literal
 * - Minimal string escaping per RFC 8785
 */
export function canonicalizeV2(obj: unknown): string {
  assertNoFloats(obj);
  return jcsSerialize(obj);
}

function jcsSerialize(value: unknown): string {
  if (value === null) return 'null';
  if (value === undefined) return 'null';

  switch (typeof value) {
    case 'boolean':
      return value ? 'true' : 'false';
    case 'number':
      return jcsNumber(value);
    case 'string':
      return JSON.stringify(value);
    case 'object':
      if (Array.isArray(value)) {
        return '[' + value.map(jcsSerialize).join(',') + ']';
      }
      return jcsObject(value as Record<string, unknown>);
    default:
      throw new Error(`Unsupported type for canonicalization: ${typeof value}`);
  }
}

function jcsNumber(n: number): string {
  if (!Number.isFinite(n)) {
    throw new Error(`Non-finite number in canonicalization: ${n}`);
  }
  // RFC 8785: integers must be serialized without decimal point
  if (Number.isInteger(n)) {
    return n.toString();
  }
  throw new Error(
    `Float value prohibited in DCP v2.0: ${n}`,
  );
}

function jcsObject(obj: Record<string, unknown>): string {
  const keys = Object.keys(obj).sort();
  const pairs: string[] = [];
  for (const key of keys) {
    const val = obj[key];
    if (val === undefined) continue;
    pairs.push(JSON.stringify(key) + ':' + jcsSerialize(val));
  }
  return '{' + pairs.join(',') + '}';
}

File: sdks/python/dcp_ai/v2/canonicalize.py
"""
DCP v2.0 canonicalization — profile `dcp-jcs-v1`.

Strict subset of RFC 8785 (JCS) with the cases JCS leaves
implementation-defined explicitly pinned. See
``spec/CANONICALIZATION_PROFILE.md`` for the normative reference.

Numeric rule: a number is valid iff its value, post-parse, is a finite
integer. ``1.0``, ``1.00``, ``1e2`` are accepted and normalised to
their integer form; ``0.1``, ``NaN``, ``±inf`` are rejected.
"""

from __future__ import annotations

import json
import math
from typing import Any


def assert_no_floats(value: Any, path: str = "$") -> None:
    """Reject any number whose value is not a finite integer.

    Accepts ``int`` and ``float`` instances whose value has no fractional
    part (``1.0``, ``100.0``). Rejects floats with fractional value
    (``0.1``), NaN, and infinities. Recurses into dicts and lists.

    The function preserves its historical name for API compatibility
    even though under profile ``dcp-jcs-v1`` it now accepts integer-
    valued floats. The wire format JSON parser cannot distinguish
    ``1`` from ``1.0`` after parse — pinning the rule post-parse is
    the only way the four DCP-AI SDKs converge byte-for-byte.
    """
    if isinstance(value, bool):  # bool subclasses int — skip the int branch
        return
    if isinstance(value, float):
        if not math.isfinite(value):
            raise TypeError(f"Non-finite number at {path}: {value!r}")
        if not value.is_integer():
            raise TypeError(
                f"Non-integer number at {path}: {value!r}. "
                f"DCP v2 (dcp-jcs-v1) requires integer values."
            )
        return
    if isinstance(value, dict):
        for k, v in value.items():
            assert_no_floats(v, f"{path}.{k}")
    elif isinstance(value, (list, tuple)):
        for i, v in enumerate(value):
            assert_no_floats(v, f"{path}[{i}]")


def _normalize_integer_floats(value: Any) -> Any:
    """Convert integer-valued floats to ``int`` ahead of ``json.dumps``.

    ``json.dumps(1.0)`` produces ``"1.0"`` and would diverge from the
    other SDKs' output. Coercing to ``int`` first guarantees the
    profile's "no decimal point, no exponent" output for any number
    whose value is a finite integer.
    """
    if isinstance(value, bool):
        return value
    if isinstance(value, float):
        # `assert_no_floats` already rejected the non-integer cases.
        return int(value)
    if isinstance(value, dict):
        return {k: _normalize_integer_floats(v) for k, v in value.items()}
    if isinstance(value, (list, tuple)):
        return [_normalize_integer_floats(v) for v in value]
    return value


def canonicalize_v2(obj: Any) -> str:
    """Canonical JSON serialisation per profile ``dcp-jcs-v1``.

    - Object keys sorted lexicographically by Unicode code point.
    - Compact form (no whitespace).
    - Integer-only numbers, decided post-parse — integer-valued floats
      are normalised, fractional floats are rejected.
    - ``None`` preserved as JSON ``null``.

    Raises ``TypeError`` if the tree contains a non-integer or
    non-finite number.
    """
    assert_no_floats(obj)
    normalised = _normalize_integer_floats(obj)
    return json.dumps(normalised, sort_keys=True, separators=(",", ":"), ensure_ascii=False)
